torch/_decomp/__init__.py

```diff
@@ -0,0 +1,76 @@
+import torch
+import torch._ops
+from typing import Callable, Union, Dict, Sequence
+from torch.utils._pytree import tree_map
+
+__all__ = ["decomposition_table", "register_decomposition", "get_decompositions"]
+
+decomposition_table = {}
+
+
+def register_decomposition(aten_op, registry=None):
+    """
+    A decorator to register a function as a decomposition to the Python
+    decomposition table.  Use it like this::
+
+        @register_decomposition(torch.ops.aten.clamp_min)
+        def clamp_min(x):
+            return torch.clamp(self, min=min)
+
+    If you are writing a new decomposition, consider contributing it
+    directly to PyTorch in torch._decomp.decompositions.
+
+    This API is experimental; we are almost certainly going to extend
+    the API when we make decompositions eligible for use in transforms (e.g.,
+    autograd) and not just backend tracing, where we then need to know if a
+    decomposition can be used to simulate a transform.
+    """
+    def decomposition_decorator(f):
+        nonlocal registry
+        if registry is None:
+            registry = decomposition_table
+
+        def add_op_to_table(aten_op):
+            # Converts aten.foo to aten.foo.default
+            # Done so I can be lazy and not write default on all of these ops
+            if not isinstance(aten_op, torch._ops.OpOverload):
```

Comment:
nit: Could this be confusing? I'm worried someone can interpret aten.foo as "this is the decomposition for all overloads of aten.foo"

Reply:
I'm OK with changing this, I was mostly trying to keep in line with the old code as much as possible. @Chillee any thoughts?

Reply:
hmmm.... I think it's potentially fine.

I wouldn't be massively opposed to forcing an explicit use of `default` on every decomposition - just makes things a bit uglier.

Reply:
I feel there is a more fundamental API design question we should resolve one way or another, which is whether or not it's OK to say torch.ops.aten.foo when we mean torch.ops.aten.foo.default. I know @anjali411 hates it. This area generally seems like an extreme attractive nuisance in the API right now. If we're going to keep the conversion, we should commit to "whenever I get an OpOverloadPacket where I expect an OpOverload, I will grab the default overload" and apply it consistently in all our torch.ops handling APIs.

Reply:
> whether or not it's OK to say torch.ops.aten.foo when we mean torch.ops.aten.foo.default

I don't think it's a good idea. Not every operator is guaranteed to have a default overload. And in many cases like add, default overload is probably not what you want since it corresponds to the following op: `aten::add(Scalar a, Scalar b) -> (Scalar)`

On a more high level, it seems bad to infer the packet as the default overload by default because calling those two operators on the same set of arguments can lead to different result (because of jit overload resolution). 

Does it make sense to register decompositions for the packets instead of the overloads? To put it another way, do we plan to register different decompositions for different overloads?

Reply:
Hypothetically, you could register a decomp for a packet, in which case you are saying "I will do overload resolution myself", but for the most part I expect people to register decomps for specific overloads. It's quite a bit nicer.